### DISCUSSION TITLE
Type hints for msgpack

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1089] [Q&A] [answered]
Hello. The code snippet below is to serialize/deserialize python objects including numpy arrays using msgpack (in combination with msgpack_numpy):
What is the type hint(s) for the parameter value in the function msgpack_serialize(value) and msgpack_deserialize(value) ?
How to write the type hints for 'a_list' and ['this', 'that', 'other'] in the write function somedb() ?
How to write the type hints for 'an_array' and np.random.random(1280) in the write function somedb() ?
Do the parameters in the somedb() functions need type hints?

This for Python 3.9+. Thank-you
```
import numpy as np
from msgpack import packb, unpackb
import msgpack_numpy as m
m.patch()

def msgpack_serialize(value):
    return packb(value, default=m.encode)

def msgpack_deserialize(value):
    return unpackb(value, object_hook=m.decode)

with somedb('mydb':str, 'w':str) as db:
    db.write('a_list', msgpack_serialize(['this', 'that', 'other']))
    db.write('an_array', msgpack_serialize(np.random.random(1280)))

with somedb('mydb':str, 'r':str) as db:
    this_list: list
    this_array: npt.ArrayLike
    this_list = msgpack_deserialize(db.read('a_list'))
    this_array = msgpack_deserialize(db.read('an_array'))
```

**Comment by srittau** [comment, marked as answer]
I'm not familiar with numpy or msgpack, but I'll try to answer your questions as well as I can.

The `value` arguments to `msgpack_serialize()` and `msgpack_deserialize()` are just passed on to `msgpack.packb()` and `unpackb()`, respectively. While `msgpack` has no type hints, at least `unpackb()` could probably simply be annotated using `bytes`: `def msgpack_deserialize(value: bytes):`. I couldn't find the definition of `packb()`, so I don't know how to best annotate it.

In the calls to `somedb()` you are trying to annotate the passed arguments. This doesn't work and is unnecessary. Just pass the value:

```python
with somedb('mydb', 'w') as db:
    ...
```

As I assume that `packb()` returns `bytes`, you could annotate the return type of `msgpack_serialize()` as `bytes` as well. `unpackb()` and consequently `msgpack_deserialize()` will probably need to return `Any` (but I might be mistaken). In this case, you can use annotations like this:

```python
with somedb('mydb':str, 'r':str) as db:
    this_list: list = msgpack_deserialize(db.read('a_list'))
    this_array: npt.ArrayLike = msgpack_deserialize(db.read('an_array'))
```

**Reply by dineshbvadhia** [reply]
`value `in `unpackb(value: bytes, object_hook=m.decode)` can be annotated as `bytes `.

`value` in `packb(value, default=m.encode)` can take the python objects lists, dicts, others and numpy arrays. So, is the annotation `value: Any` ?

**Reply by srittau** [reply]
You could annotate it with a union of all accepted types, but I think using `Any` is the easiest.

**Reply by dineshbvadhia** [reply]
Using `Union` will be better as I can add new object annotations in the future. This has been very informative. Thank-you.
